docs/proposals/2841-flux-hpc/README.md

```diff
@@ -0,0 +1,275 @@
+4.  **Introduce an Extensible API Policy.** Define a new `hpcPolicy` that is generic enough to support other HPC workload managers in the future (e.g., Slurm, LSF), with Flux as the initial, reference implementation.
+
+### Non-Goals
+
+1.  **Replace JobSet:** This proposal extends and enhances `JobSet`, not replaces it. `JobSet` remains the core abstraction for managing the group of pods.
+2.  **Support Additional HPC Schedulers Immediately:** The initial implementation will focus exclusively on Flux Framework. Support for other managers can be added later under the same `hpcPolicy` API by respective parties that have interest.
+3.  **Re-implement the MPI Operator:** This proposal provides an alternative to the MPI Operator's launcher/worker model by leveraging Flux's native capabilities, rather than replicating its logic.
+
+## Proposal
+
+The core of this proposal is to introduce a new Kubeflow Trainer plugin named `Flux`. This plugin will implement the `ComponentBuilderPlugin` interface to modify the `JobSet` specification generated for a `TrainJob`. The mechanism for creating the Flux MiniCluster is dynamic and non-intrusive to the user's container image:
+
+1.  **API Trigger**: The user enables the feature by defining an `hpcPolicy` in their `TrainJob` runtime specification and setting the `manager` to `"flux"`.
+2.  **Plugin Activation**: The Kubeflow Trainer controller invokes the `Flux` plugin's `Build` method.
+3.  **JobSet Modification**: The plugin modifies the `JobSet` specification before it is created:
+    *   An **`initContainer`** is added to the "trainer" replicated job. This container uses a pre-built "flux-view" image containing a Spack installation of Flux.
+    * A **pod affinity** is added that enforces a soft requirement to schedule one pod per node to support Flux controlling the mapping of all node resources. An optional **node affinity** can enforce that the MiniCluster pods are only scheduled to specific nodes.
+    *   A **shared memory mount** that ensures the pod can utilize all shared memory on the node (or a reasonable subset). By default most container runtimes will mount only 64M, and this can negatively impact MPI performance.
+    *   **Shared `emptyDir` Volumes** are mounted into both the `initContainer` and the main application container to move the Flux install from the initContainer to the application container. The `initContainer` copies the Flux installation and necessary software from its own image into these shared volumes, and generates configuration for the cluster based on the user-preferences provided.
+    *   A **ConfigMap** is generated containing two scripts: `init.sh` (for the init container) and `entrypoint.sh` (for the main container). This ConfigMap is mounted into the pods.
```

Comment:
Is the main responsibility of ConfigMap to place these scripts into containers file-system, so they can execute it? I am wondering why you don't want to directly define this script in the container command/args?

Reply:
It would be way too long to define as an entrypoint, and probably lead to many bugs. It also allows the user to define custom commands at different steps in the setup to run (init, pre, post, etc). It works as a template with the CRD.